### ISSUE TITLE
Validation criteria: `verify_strict` might be accepting non-canonical `R`

### ISSUE THREAD

**Issue by rozbb** [dalek-cryptography/ed25519-dalek#267]
It [appears](https://github.com/dalek-cryptography/ed25519-dalek/blob/6d9bbd323edfce04f600427571e90afd86f52939/src/verifying.rs#L307) that `verify_strict` might permit [non-canonical `R`](https://hdevalence.ca/blog/2020-10-04-its-25519am), while `verify` does not. Notice the difference in the code is that `verify_strict` does a point comparison, while `verify` does a byte comparison.

This is pretty subtle, and I think it's not detected by our validation tests because the validation tests never do a non-canonical `R` that isn't also low-order. Three things I wanna do:
1. Confirm that this behavior is happening
2. Update `verify_strict` to reject non-canonical encoding
3. Write a regression test similar to what [zebra has](https://github.com/ZcashFoundation/ed25519-zebra/blob/612e51af2e7eaa228f9a08ad38b0b0660e510d9e/tests/util/mod.rs#L81-L155)

This is technically breaking, but I consider it a bugfix rather than a validation change. Nobody would reasonably expect `verify_strict` to be more permissive than `verify` on any axis.

**Comment by rozbb** [comment]
Ok I think this is a non-issue. Yes, it may be the case that `verify_strict` accepts prime-order non-canonical `R`, but the reason it's not an issue is because it's extremely difficult to generate a signature with such an `R`. That's because the ability to produce such a signature is the same as proving knowledge of of `dlog(R)`, which is probably hard.

My proposed solution: let's just make this explicit, and do verification by byte comparison in all `verify_*` functions. This will simplify the code and only change behavior in a computationally infeasible edge case.

cc @tarcieri

**Comment by tarcieri** [comment]
> My proposed solution: let's just make this explicit, and do verification by byte comparison in all verify_* functions. 

Agree with this, and I wonder if you could potentially wrap it up a little more, perhaps a function that calls your proposed `recompute_r` function and compares it with an expected `R` in a single place which can be reused?

**Comment by rozbb** [comment]
My thought was to have `recompute_r` just return the compressed form, and then all the callers have no choice but to do a byte comparison. This looks nice enough imo, since everything all the verify functions look like `expected_R == computed_R`.

But #251 needs to land before that happens.

**Comment by tarcieri** [comment]
ed25519-dalek has moved to: https://github.com/dalek-cryptography/curve25519-dalek/tree/main/ed25519-dalek

If you feel this issue has not yet been addressed, please request this issue be moved to that repo by leaving a comment, or open a new issue on that repo.
